In [13]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data = pd.read_csv("../Resources/lending_data.csv")


# Review the DataFrame
lending_df = pd.DataFrame(lending_data)
lending_df.head(10)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   
5    10100.0          7.438            50600        0.407115                4   
6    10300.0          7.490            51100        0.412916                4   
7     8800.0          6.857            45100        0.334812                3   
8     9300.0          7.096            47400        0.367089                3   
9     9700.0          7.248            48800        0.385246                4   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  
5                 1       20600            0  
6                 1       21100            0  
7                 0       15100            0  
8                 0       17400            0  
9                 0       18800            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features
y = lending_df['loan_status']
X = lending_df.drop(columns = 'loan_status')
                    

In [4]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
X[:5]

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [6]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split


# Split the data using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y,  random_state=1)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [7]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Assign a random_state of 1 to the function
model = LogisticRegression(random_state=1)
model

# Fit the model using training data
logreg = model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [8]:
# Make a prediction using the testing data
predictions = logreg.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [19]:
# Generate a confusion matrix for the model
matrix = confusion_matrix(y_test, predictions)
matrix = pd.DataFrame(matrix,
                     index = ['Actual Healthy Loan', 'Actual High-Risk Loan'],
                     columns = ['Predicted Healthy Loan', 'Predict High-Risk Loan'])
matrix

Predicted Healthy Loan  Predict High-Risk Loan
Actual Healthy Loan                     18663                     102
Actual High-Risk Loan                      56                     563

In [22]:
# Print the classification report for the model
target_names  = ["Healthy Loan (0)", "High-Risk Loan (1)"]
report = classification_report(y_test, predictions, target_names  = target_names)

# Print the training classification report
print(report)

                    precision    recall  f1-score   support

  Healthy Loan (0)       1.00      0.99      1.00     18765
High-Risk Loan (1)       0.85      0.91      0.88       619

          accuracy                           0.99     19384
         macro avg       0.92      0.95      0.94     19384
      weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
For Healthy Loans (0), this logistic regression model demonstrates outstanding performance, having perfect precision and f1-score of 1.00. This means it accurately identifies nearly all healthy loans.

When it comes to High-Risk Loans (1), the logistic regression model performs well with a precision of 0.85, It correctly predicts high-risk loans 85% of the time. A recall of 0.91 shows it detects 91% of all true high-risk loans. The F1-score of 0.88 indicates a strong balance between precision and recall for this category.

Overall, the logistic regression model has a high accuracy rate of 0.99, correctly classifying 99% of loans. The macro average F1-score of 0.94 suggests solid performance across both loan types, while the weighted average F1-score of 0.99 underscores the model's excellent performance.